ocaml/xapi/extauth_ad.ml

```diff
@@ -0,0 +1,74 @@
+(*
+ * Copyright (C) 2006-2009 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+(**
+ * @AD group Access Control
+*)
+
+open Db_actions
+
+module D = Debug.Make (struct let name = "extauth_ad" end)
+
+open D
+
+exception Unknown_Ad_type of string
+
+type ad_type = Pbis | Winbind
+
+module Ad_type = struct
+  type t
+
+  let of_string = function
+    | "pbis" ->
+        Pbis
+    | "winbind" ->
+        Winbind
+    | _ as at ->
+        debug "Unknow ad type of %s" at ;
+        raise (Unknown_Ad_type at)
+end
+
+module AD = struct
+  let start_backend_daemon ~wait_until_success = function
+    | Pbis ->
+        Extauth_plugin_ADpbis.Lwsmd.start ~wait_until_success ~timeout:5.
+    | Winbind ->
+        Extauth_plugin_ADwinbind.Winbind.start ~wait_until_success ~timeout:5.
+
+  let stop_backend_daemon ~wait_until_success = function
+    | Pbis ->
+        Extauth_plugin_ADpbis.Lwsmd.stop ~wait_until_success ~timeout:3.
+    | Winbind ->
+        Extauth_plugin_ADwinbind.Winbind.stop ~wait_until_success ~timeout:3.
+
+  let init_service ~__context = function
+    | Pbis ->
+        Extauth_plugin_ADpbis.Lwsmd.init_service ~__context
+    | Winbind ->
+        Extauth_plugin_ADwinbind.Winbind.init_service ~__context
+end
+
+let start_backend_daemon ~wait_until_success =
```

Comment:
Suppose we were using PBIs, and then xapi crashed, but before rebooting, we changed the AD backend to winbind. Would we then have both daemons running? 

I doubt we want this - maybe in this function we could make we sure we shutdown any other daemons before starting the current backend

Reply:
change the AD backend is only a temporary emergency  solution, for your case, we would expect customer (or script) to
- stop xapi (skip if xaip crash)
- stop winbind daemon
- change ad_backend in xapi.conf
- start xapi 

If we coding that in xapi,  and always stop pbis before start winbind,  this will always introduce extra failures in xapi logs when it try to stop PBIS.  As the PBIS baries and services is not installed at all.

Reply:
I think we can tell systemd itself to not start both daemons at once, with a `Conflicts` directive, e.g. `Conflicts=winbind.service`. This is probably a simpler solution than having xapi manage this

Reply:
Use `Conflicts` directive sounds a clean solution to me,  but 
- if we put this `Conflicts` directive in PBIS (lwsmd), then we need to update PBIS before we remove PBIS,  (we will need to include update PBIS binaries to remove PBIS0, sounds strange!
- if we put this `Conflicts` directive in winbind, we have to setup new patchqueu in winbind, and patch winbind for this.

I do not think it worth a patchqueu just for this temporary solution (PBIS does not live long as we will finally remove it),   so I create ticket in CoP for this and future improvements.
but for now, let just make sure we will not start both by scripts to change backends.